### DISCUSSION TITLE
Direction for the Stack

### DISCUSSION THREAD

**Discussion by Rykuno** [Rykuno/fullstack-example#6] [Announcements]
Hello friends. I made this boilerplate public so I could link individuals on Reddit/Discord to the solutions I have found great when building more complex fullstack apps with the best DX I've experienced. I'd say its a good middle ground between just using Svelte/Next api endpoints and something like NestJS. Although, if you need multiple microservices or some CQRS architecture, just use a monorepo and NestJS. 

 I've been receiving too many emails and reddit messages to respond to all of them reasonably, so I've recently opened the discussion/issues here. 

I have a private repo in the works with a full documentation site, cleaner more flexible auth, and a one click deploy to both railway and render.com. I'll also create a Discord for you guys to discuss architecture and help each other as I have a fulltime job, family, side-businesses, and a life to waste in MMORPGS.

## Manefesto 
The point of this is to spin up MVP's or weekend projects in a scalable, maintainable, and extensible manner. ***THERE ARE 0 EXTERNAL SAAS YOU NEED(other than email, which I use SES)***. The thing that pisses me off the most is when i need to manage multiple accounts, configs, payments, and signups. Fuck that. Let me emulate my entire stack locally and deploy with 1 click through my choice of hosting provider. 

## FAQ
To address a few of the questions I get,

1) Storage - My storage solution is a self hosted minIO instance. I do not use Cloudflare R2 or AWS S3 simply because I don't need to store petabytes of data and I prefer a cheaper, faster, more modular solution that can be entirely emulated locally. IF I need to switch, MinIO uses an S3 adapter and both R2 and S3 make it REALLY REALLY easy to migrate to their platforms. You guys know how cheap disk storage is though, right?

2) Inversion of Control - I believe in testing. I work on projects either myself or with a small team. IoC allows me to test the things that keep me up at night easily, like auth, payments, registration, or important jobs. It also helps provide a modular structure that promotes decoupling and maintainability. If you ever come back to a project after a week and don't recognize what you wrote, you need structure. 

3) "Can you add X Libaray" - No, this is a bare bones template that gives you auth, rate limiting, queues/jobs, and a structure to build on-top of. The point of this template is for you to fork it and do as you please. Its free. Use what I have for inspiration and either mold mine or make your own. 

4) Folder/File Architecture - The folder architecture is based off "Technical Architecture". Everything is organized into their programatic function. I find this is a very good way to start projects off and eventually evolve into VSA or Onion/CLEAN later on. Because all the files are named according to their function(eg `iam.controller.ts` and `iam.service.ts`) this makes it easy to move them around without conflict. Move things around until they feel right or follow your favorite architecture pattern.

**Comment by luxus** [comment]
looking forward to get some ideas how you did some integrations. i'm not really into the "one click deployment" stuff, but looking forward to see it as a reference for lucia and passkey intigration with sveltekit and hono

**Reply by Rykuno** [reply]
> looking forward to get some ideas how you did some integrations. i'm not really into the "one click deployment" stuff, but looking forward to see it as a reference for lucia and passkey intigration with sveltekit and hono

The next version does not have passkey/oauth built in, but rather a simple step by step guide on how to do it. 

I'm curious, what is the issue with "one click deployment"? I can clone the repo atm and have the entire thing hosted in production instantly without signing up for any services other than an email provider.

**Comment by fres-sudo** [comment]
Hello Rykuno, I really appreciate your work I've been using your stack for some times now and I'm learning a lot by using it, I really enojoy the DX and all the architecture stuff.

It would be awesome to see also a OAuth integration! I'm open if you need some contributors, I can try my  best to implement it!

**Reply by Rykuno** [reply]
> Hello Rykuno, I really appreciate your work I've been using your stack for some times now and I'm learning a lot by using it, I really enojoy the DX and all the architecture stuff.
> 
> It would be awesome to see also a OAuth integration! I'm open if you need some contributors, I can try my best to implement it!

OAuth integration isn't built in, but rather I'll have a simple step by step guide on how to implement it. Trying to keep it modular/minimal.

**Reply by fres-sudo** [reply]
Oh cool! I understand the reason and I agree with you. If you wanna some help with anything I'm glad to help you out.
Have a great day!

**Comment by itisyb** [comment]
I can work directly with Supabase SDK 
Add caching and stuff

do you think tanstack is a good addition for something like that?
